feat: add documentation and UI integration for Before User Created hook

cstockton opened the pull request:
Add documentation and UI integration for Before User Created hook ([supabase/auth#2034](https://github.com/supabase/auth/pull/2034)):
- Added new guide: `before-user-created-hook.mdx`, documenting usage, inputs, outputs, and multiple complete examples for HTTP and SQL.
- Updated navigation menu and hook index table to include the new hook.
- Hook is now listed as available on Free and Pro plans.

Studio changes:
- Registered `Before User Created` hook in `hooks.constants.ts` with proper title, slug, and configuration keys.
- Enabled Docs button linking to hook guide in both `CreateHookSheet` and `HookCard` components.
- Extended hook listing page to support docs access via `secondaryActions`.

This hook allows developers to intercept and reject user creation across all flows (email/password, OAuth, SSO, SAML, anonymous) - enabling custom signup logic like domain allowlists, geofencing, or account gating.

Commit:
feat: add documentation and UI integration for Before User Created hook

- Added new guide: `before-user-created-hook.mdx`, documenting usage, inputs,
  outputs, and multiple complete examples for HTTP and SQL.
- Updated navigation menu and hook index table to include the new hook.
- Hook is now listed as available on Free and Pro plans.

Studio changes:
- Registered `Before User Created` hook in `hooks.constants.ts` with proper
  title, slug, and configuration keys.
- Enabled Docs button linking to hook guide in both `CreateHookSheet` and `HookCard` components.
- Extended hook listing page to support docs access via `secondaryActions`.

This hook allows developers to intercept and reject user creation across all
flows (email/password, OAuth, SSO, SAML, anonymous) - enabling custom signup
logic like domain allowlists, geofencing, or account gating.

hf commented on apps/docs/content/guides/auth/auth-hooks/before-user-created-hook.mdx:
```diff
@@ -0,0 +1,488 @@
+
+This hook runs before a new user is created. It allows developers to inspect the incoming user object and optionally reject the request. Use this to enforce custom signup policies that Supabase Auth does not handle natively - such as blocking disposable email domains, restricting access by region or IP, or requiring that users belong to a specific email domain.
+
+The hook is triggered any time a new user would be created - including traditional email signups (signUp()), OAuth flows (signIn(\{ provider \})), SSO (signInWithSSO()), IdP-initiated SAML flows, and anonymous sign-ins. Before the user is inserted into the auth.users table, the payload is passed to your configured hook endpoint for evaluation.
```
Maybe this would read better if you don't exhaustively list all of the places where it runs?

```suggestion
The hook is triggered any time a new user would be created - including traditional email signups (signUp()), OAuth flows (signIn(\{ provider \})), SSO (signInWithSSO()), IdP-initiated SAML flows, and anonymous sign-ins. Before the user is inserted into the `auth.users` table, the payload is passed to your configured hook endpoint for evaluation.
```

cstockton replied:
I've updated the text.

hf commented on apps/docs/content/guides/auth/auth-hooks/before-user-created-hook.mdx:
```diff
@@ -0,0 +1,488 @@
+
+| Field    | Type     | Description                                                                               |
+| -------- | -------- | ----------------------------------------------------------------------------------------- |
+| `header` | `object` | Metadata about the request. Includes IP address, request ID, and hook type.               |
```
Didn't we change this to `metadata`?

```suggestion
| `metadata` | `object` | Metadata about the request. Includes IP address, request ID, and hook type.               |
```

hf commented on apps/docs/content/guides/auth/auth-hooks/before-user-created-hook.mdx:
```diff
@@ -0,0 +1,488 @@
+| Field    | Type     | Description                                                                               |
+| -------- | -------- | ----------------------------------------------------------------------------------------- |
+| `header` | `object` | Metadata about the request. Includes IP address, request ID, and hook type.               |
+| `user`   | `object` | The user record that is about to be created. Matches the shape of the `auth.users` table. |
```
Add additional information that this user **does not exist in the database** when the function is ran, so you will not be able to `select ... from auth.users where id = <id-from-hook>` nor use Supabase Auth Admin API to get the user.

cstockton replied:
Added a note on this.

hf commented on apps/docs/content/guides/auth/auth-hooks/before-user-created-hook.mdx:
````diff
@@ -0,0 +1,488 @@
+
+```json
+{
+  "header": {
````
Isn't this metadata? https://github.com/supabase/auth/blob/master/internal/hooks/v0hooks/v0hooks.go#L58

cstockton replied:
Fixed

hf commented on apps/docs/content/guides/auth/auth-hooks/before-user-created-hook.mdx:
```diff
@@ -0,0 +1,488 @@
+| -------- | --------- | ----------------------------------------------------------------------------------------------------- |
+| `error`  | `object`  | (Optional) Return this to reject the signup. Includes a code, message, and optional HTTP status code. |
+
+Returning an empty object with a `200` or `204` status code allows the request to proceed. Returning a JSON response with an `error` object and a `4xx` status code blocks the request and propagates the error message to the client. See the [error handling documentation](/docs/guides/auth/auth-hooks#error-handling) for more details.
```
Returning `null` (SQL equivalent of HTTP 204) should also be OK.

hf reviewed (commented):
Looks good. @charislam could we get your eyes on this as well?

Commit:
fix: add subnet to mdx lint allowed list

Commit:
fix: incorporate feedback and spelling changes

cstockton commented on apps/docs/content/guides/auth/auth-hooks/before-user-created-hook.mdx:
```diff
@@ -0,0 +1,488 @@
+* Detecting that a user is about to be created
+* Providing the ip address in the `header.ip_address` field
+
+This example uses PostgreSQL’s built-in `inet` and `<<` CIDR operators:
```
👍

charislam reviewed (approved):
only read the docs changes, LGTM

Commit:
feat: add allow/deny domain and CIDR examples to before-user-created hook

This commit enhances the documentation for the before-user-created auth hook:
- Adds production-ready SQL examples for blocking or allowing signups by email domain (signup_email_domains) and IP address or CIDR (signup_networks)
- Introduces enum-based classification (allow vs deny) for both domain and network restrictions
- Adds full migration-style snippets for easy copy/paste into Supabase SQL Editor
- Includes updated HTTP examples that delegate to Postgres functions via Supabase client rpc()
- Adds environment setup guidance and edge function scaffolding

Commit:
fix: pnpm format

cstockton committed:
Merge branch 'master' into cs/master-add-before-user-created-hook
